### ISSUE TITLE
sqlite3.connect, return type and factory argument

### ISSUE THREAD

**Issue by shawnbrown** [python/typeshed#11558]
When using the `sqlite3.connect()` function with its  `factory` argument, type checking fails when I believe it should pass.

For the following tests, I used:

```text
$ python --version
Python 3.12.2

$ python -m mypy --version
mypy 1.9.0 (compiled: yes)
```

### Type checking fails when I believe it should pass

```python
# check_return_type.py

import sqlite3

class MyConnection(sqlite3.Connection):
    """My custom SQLite connection subclass."""

con: MyConnection = sqlite3.connect(':memory:', factory=MyConnection)
assert isinstance(con, MyConnection)
print('OK!')
```

Type checking fails:

```text
$ python -m mypy check_return_type.py
check_return_type.py:8: error: Incompatible types in assignment (expression has type "Connection", variable has type "MyConnection")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

But running this script works as expected:

```text
$ python check_return_type.py 
OK!
```

I've only started using generics recently so I don't feel qualified to submit a patch for this myself.

### Type checking passes when I believe it should fail

```python
# check_optional.py

import sqlite3

con: sqlite3.Connection = sqlite3.connect(':memory:', factory=None)
con.close()
```

And here, type checking passes:

```text
$ python -m mypy check_optional.py 
Success: no issues found in 1 source file
```

But running this script fails:

```text
$ python check_optional.py 
Traceback (most recent call last):
  File "~/Projects/typeshed-testing/check_optional.py", line 4, in <module>
    con: sqlite3.Connection = sqlite3.connect(':memory:', factory=None)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not callable
```

Admittedly this second issue could be considered semantic, ~~but the `sqlite3.connect()` function's `factory` argument  was introduced in Python 3.10 and--at least according to the docs--it~~ has never been implemented as `Optional` (but always used `sqlite3.Connection` as a default value).

edit: The `factory` argument was present but just undocumented until Python 3.10.

**Comment by srittau** [comment]
We don't handle `factory` specially at all in the stubs. We need to add appropriate overloads. PR welcome!

**Comment by shawnbrown** [comment]
I'm not sure I'm the best person to submit a PR for this. But if you think the changes I made in my fork are acceptable (https://github.com/shawnbrown/typeshed/commit/2e51ff1dff8daec7039cba9c77d9940db7544e79), then I'll submit them.

I wasn't sure how to resolve this with overload like you mentioned so I mimicked how the existing code handled the `Cursor` factory using the `_CursorT` generic.

**Comment by srittau** [comment]
@shawnbrown I've left a few comments on that commit. I think with these changes, it would be a good basis for a PR. I especially like the extensive tests you added.

**Comment by shawnbrown** [comment]
Thanks for your comments. I've updated my changes to address the things you mentioned (please see https://github.com/shawnbrown/typeshed/commit/6eaf1b693a472643bcae63bbc93c1db42e0ff26c).

**Comment by JelleZijlstra** [comment]
Thanks! I'd recommend you open a PR so it's easier to review; PRs are cheap. You can leave it in draft if you prefer.

**Comment by shawnbrown** [comment]
Oops. I forgot to close this. The is resolved and the related pull request was merged a couple of days ago (#11571). Thanks for everyone's help on this.
